### PR TITLE
Improvements to the Random class.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
* Replaced the RNG algorithm with the smallest and most secure [Permuted Cogruential Generator(PCG)] (http://www.pcg-random.org/), as implemented [here](https://github.com/imneme/pcg-c-basic/blob/master/pcg_basic.c).
* Added methods that seed the class with longer seeds. The original seeding methods are retained.
* Formatted the source code file for the Random class in order to improve readability and to maintain a consistent code style across the file.

### CODE REVIEW

**Comment by danmosemsft** [comment]
@teo-tsirpanis thanks for the contribution! This probably should have some longer discussion which is better in an issue. Please close this PR for now and open an issue linked to it so we can talk about it. We will need to consider compatibility issues, randomness quality and also performance. You will want to make some measurements comparing the old and new implementations to check this is at least no slower.

File: src/mscorlib/src/System/Random.cs [hunk 1]
```diff
@@ -10,205 +10,192 @@
...
+
+        private int InternalSample()
+        {
+            var oldstate = state;
+            state = oldstate * MAGIC + inc;
+            var xorshifted = (uint)(((oldstate >> 18) ^ oldstate) >> 27);
+            var rot = (uint)oldstate >> 59;
+            return (int) ((xorshifted >> (byte)rot) | (xorshifted << (byte)((-rot) & 31)));
+        }
+
```
Comment (redknightlois) [inline src/mscorlib/src/System/Random.cs]: From the point of view of performance, make sure the the JIT is inlining InternalSample or request for the JIT to aggresively inline it.

Comment (teo-tsirpanis) [inline src/mscorlib/src/System/Random.cs]: And I do that by adding `[MethodImplOptions.AggressiveInlining]` in `InternalSample`?

**Comment by SaxxonPike** [comment]
I understand this would be a breaking change for a program relies on the specific deterministic behavior of Random based on a seed (for example, game replays). Would you suggest with something like this, not using the built in Random class for this purpose anymore?

**Comment by teo-tsirpanis** [comment]
Of course not. My only objection is that the existing algorithm is obscure and the new algorithm is smaller and more difficult to predict. The `Random` class will have the same use as it always had. Now, regarding backwards compatibility, we could implement an `AppContext` switch to either enable or disable the new RNG.

**Comment by jkotas** [comment]
The discussion on this can continue in dotnet/corefx#12746. Even though the built-in random algorithm is not perfect, changing it would very likely cause many breakages accross the ecosystem - it is unlikely we can accept changes in it.

**Pull request closed without merging**